cognite_toolkit/_cdf_tk/utils/fileio/_readers.py

```diff
@@ -75,34 +79,130 @@ class YMLReader(YAMLBaseReader):
-        except ValueError:
-            pass
-        if value.lower() in ("true", "false"):
-            return value.lower() == "true"
-        return value
+        return infer_data_type_from_value(value, "Json")[1]
+
+    @classmethod
+    def sniff_schema(cls, input_file: Path, sniff_rows: int) -> list[SchemaColumn]:
+        """Sniff the schema from the first `sniff_rows` rows of the CSV file."""
+        if sniff_rows <= 0:
+            raise ValueError("`sniff_rows` must be a positive integer.")
+
+        with input_file.open("r", encoding="utf-8") as file:
+            reader = csv.DictReader(file)
+            sample_rows: list[dict[str, str]] = []
+            for no, row in enumerate(reader):
+                if no >= sniff_rows:
+                    break
+                sample_rows.append(row)
+
+            if not sample_rows:
+                raise ValueError(f"No data found in the file: {input_file}")
+
+            columns = list(sample_rows[0].keys())
+            if len(columns) != len(set(columns)):
+                raise ToolkitValueError(
+                    f"Duplicate column names found in the first {sniff_rows} rows of the file: {input_file.as_posix()!r}. "
+                    "Please ensure that all column names are unique."
+                )
+            schema = []
+            for column_name in columns:
+                sample_values = [row[column_name] for row in sample_rows if column_name in row]
+                if not sample_values:
+                    column = SchemaColumn(name=column_name, type="string")
+                else:
+                    data_types = Counter(
+                        infer_data_type_from_value(value, dtype="Json")[0]
+                        for value in sample_values
+                        if value is not None
+                    )
+                    if not data_types:
+                        inferred_type = "string"
+                    else:
+                        inferred_type = data_types.most_common()[0][0]
+                    # Json dtype is a subset of Datatype that SchemaColumn accepts
+                    column = SchemaColumn(name=column_name, type=inferred_type)  # type: ignore[arg-type]
+                schema.append(column)
+        return schema
+
+    def _read_chunks_from_file(self, file: TextIOWrapper) -> Iterator[dict[str, JsonVal]]:
+        for row_no, row in enumerate(csv.DictReader(file), start=1):
+            parsed: dict[str, JsonVal] = {}
+            for key, value in row.items():
+                if value == "":
+                    parsed[key] = None
+                    continue
+                try:
+                    parsed[key] = self.parse_function_by_column[key](value)
+                except ValueError as e:
+                    parsed[key] = None
+                    if self.keep_failed_cells:
+                        self.failed_cell.append(FailedParsing(row=row_no, column=key, value=value, error=str(e)))
```

Comment:
failed_cell is not cleared anywhere, so the errors are going to get appended everytime reader is called?